### PR TITLE
dts: bindings: add support for the flash driver name

### PR DESCRIPTION
**Pull request by nvlsianpu**
So far, DT did not support the flash driver name.
Any flash-controller should have the appropriate
flash driver that should be identified by its name.

This path adds generic support for extract description
from the flash-controller node,
adds implementation of this property for all nrf5x targets
and removes previous FALSH_DRIVER_NAME macros form
nrf5x soc.h files.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

### CODE REVIEW

**Comment by nvlsianpu** [comment]
cc @KaSroka

File: dts/bindings/flash_controller/flash-controller.yaml [hunk 1]
```diff
@@ -0,0 +1,20 @@
+---
```
Comment (carlescufi) [inline dts/bindings/flash_controller/flash-controller.yaml]: How come we didn't have this before, if flash controllers were already being used in the `.dtsi` files?

Comment (nvlsianpu) [reply dts/bindings/flash_controller/flash-controller.yaml]: Not flash controller was used - but its sub-node for flash memory description.

Comment (carlescufi) [reply dts/bindings/flash_controller/flash-controller.yaml]: Ok, understood. This looks good to me then.

**Review by carlescufi — APPROVED** [review]
I assume other platforms will add support for this in separate patches

File: boards/arm/96b_carbon_nrf51/dts.fixup [hunk 2]
```diff
@@ -2,3 +2,5 @@
 #define CONFIG_UART_NRF5_IRQ_PRI	NORDIC_NRF_UART_40002000_IRQ_0_PRIORITY
 #define CONFIG_UART_NRF5_BAUD_RATE	NORDIC_NRF_UART_40002000_CURRENT_SPEED
 #define CONFIG_UART_NRF5_NAME		NORDIC_NRF_UART_40002000_LABEL
+
+#define CONFIG_SOC_FLASH_NRF5_DEV_NAME	NRF_NRF52_FLASH_CONTROLLER_4001E000_LABEL
```
Comment (galak) [inline boards/arm/96b_carbon_nrf51/dts.fixup]: Can you also add to all the fixups:

#define FLASH_DEV_NAME CONFIG_SOC_FLASH_NRF5_DEV_NAME

File: dts/arm/nordic/nrf51822.dtsi [hunk 3]
```diff
@@ -20,6 +20,8 @@
 			#address-cells = <1>;
 			#size-cells = <1>;
 
+			label="NRF5_FLASH";
```
Comment (galak) [inline dts/arm/nordic/nrf51822.dtsi]: Seems like this should be "NRF5_FLASH_CTRL";

So we can also have a label on the flash itself if needed.

File: dts/arm/nordic/nrf52832.dtsi [hunk 4]
```diff
@@ -20,6 +20,8 @@
 			#address-cells = <1>;
 			#size-cells = <1>;
 
+			label="NRF5_FLASH";
```
Comment (galak) [inline dts/arm/nordic/nrf52832.dtsi]: Same comment about this being "NRF5_FLASH_CTRL";

**Review by galak — CHANGES_REQUESTED** [review]
Can you also fix the type in the commit message - FLASH isn’t spelled correctly.

File: dts/arm/nordic/nrf52840.dtsi [hunk 5]
```diff
@@ -20,6 +20,8 @@
 			#address-cells = <1>;
 			#size-cells = <1>;
 
+			label="NRF5_FLASH";
```
Comment (galak) [inline dts/arm/nordic/nrf52840.dtsi]: Same comment about this being "NRF5_FLASH_CTRL";

File: dts/bindings/flash_controller/flash-controller.yaml [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+description: >
+    This binding gives the base structures for all flash controller devices
+
+properties:
+    label:
+      type: string
+      category: required
+      description: Human readable string describing the device (used by Zephyr for API name)
+      generation: define
+    interrupt-names:
```
Comment (galak) [inline dts/bindings/flash_controller/flash-controller.yaml]: Why did you feel like interrupt-names, belonged here, but not interrupts or reg?

File: dts/bindings/flash_controller/flash-controller.yaml [hunk 7]
```diff
@@ -0,0 +1,20 @@
+---
+title: flash controller Base Structure
+id: flash-controller
```
Comment (galak) [inline dts/bindings/flash_controller/flash-controller.yaml]: id: soc-nv-flash-controller

File: dts/arm/nordic/nrf51822.dtsi (hunk 3 above)
Comment (nvlsianpu) [reply dts/arm/nordic/nrf51822.dtsi]: NRF5_FLASH is used to identify the nordic flash driver (along with device_get_bindings call). I would change this to something like NRF5_FLASH_NAME. ?

**Review by galak — CHANGES_REQUESTED** [review]
Some minor changes.

File: dts/bindings/flash_controller/flash-controller.yaml [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+
+    reg:
+      type: array
+      description: mmio register space
+      generation: define
+      category: required
+
+    interrupts:
+      type: array
+      category: required
```
Comment (galak) [inline dts/bindings/flash_controller/flash-controller.yaml]: Should be optional

File: arch/arm/soc/nordic_nrf5/nrf52/soc.h [hunk 9]
```diff
@@ -24,9 +24,4 @@
 #define FLASH_PAGE_ERASE_MAX_TIME_US 	89700UL
 #define FLASH_PAGE_MAX_CNT		256UL
 
-/* For IMG_MANAGER  */
```
Comment (galak) [inline arch/arm/soc/nordic_nrf5/nrf52/soc.h]: I'd rather we keep FLASH_DRIVER_NAME here, and not duplicate it in all the fixup files.  Hopefully we can get ride of it in the future.

So can we keep the comment and such and just have:

#define FLASH_DRIVER_NAME FLASH_DEV_NAME

**Comment by nvlsianpu** [comment]
@galak I think satisfy all comments

File: dts/bindings/flash_controller/nordic,nrf51-flash-controller.yaml [hunk 10]
```diff
@@ -0,0 +1,19 @@
+---
+title: Nordic NVMC
+id: nordic,nrf-nvmc
```
Comment (galak) [inline dts/bindings/flash_controller/nordic,nrf51-flash-controller.yaml]: nit, but id: should try and match the compat

id: nordic,nrf51-flash-controller

File: dts/bindings/flash_controller/nordic,nrf52-flash-controller.yaml [hunk 11]
```diff
@@ -0,0 +1,19 @@
+---
+title: Nordic NVMC
+id: nordic,nrf-nvmc
```
Comment (galak) [inline dts/bindings/flash_controller/nordic,nrf52-flash-controller.yaml]: id: nordic,nrf52-flash-controller

**Review by galak — CHANGES_REQUESTED** [review]
Is there actually any difference between the NRF51 & NRF52 flash controllers?

If not we should use the same compatible for them (and same yaml).

**Comment by nvlsianpu** [comment]
Yes - nrf51 flash controller is not the same as nrf52 flash controller. for instance nrf52 has some more register, timings are different, At this point I don't include this differences in this PR.

**Comment by galak** [comment]
> Yes - nrf51 flash controller is not the same as nrf52 flash controller. for instance nrf52 has some more register, timings are different, At this point I don't include this differences in this PR.

Ok, sounds good.  Lets just fixup the ids in the yaml

**Review by galak — CHANGES_REQUESTED** [review]
Also the commit message should be fixed, since we didn’t remove FLASH_DRIVER_NAME

**Review by galak — APPROVED** [review]

**Comment by nvlsianpu** [comment]
@anangl  examle of DT driver name extension.

**Pull request merged**